### ISSUE TITLE
Provide suggestions only for the C# language version specified in the project settings (and lower versions)

### ISSUE THREAD

**Issue by ironcev** [ironcev/Sharpen#13]
Currently all the suggestions are provided, even if the project settings are set to a C# language version that does not support a particular suggestion. E.g. if the project sets the C# language version to 5.0 the suggestions for the version 7.0 should not be provided.

**Comment by ironcev** [comment]
A lower language version used in the project is just one of the possible reasons why a suggestion cannot be applied. Type availability in the framework is another one. E.g. only some of the features of C# 8.0 are directly supported in .NET 4.8.

![image](https://user-images.githubusercontent.com/4142833/58090429-80bba400-7bc7-11e9-89d4-9538a5d80528.png)

We could even have a situation that the framework version does not support a certain type like e.g. `Range` but the project uses NuGet packages that provide the type.

On the other hand, for e.g. Default interface implementation there are no special framework types needed, but the feature will work only on the Core runtime (at the moment).

Long story short: it's complex 😄 
So in general, think well about what it really means that a particular suggestion can be applied in a particular piece of code.

(The table with language feature and framework support is taken from @miroslavpopovic's presentation that can be found in his [what-is-new-in-dotnet-core-3](https://github.com/miroslavpopovic/what-is-new-in-dotnet-core-3) repository.)

**Comment by shankyjain7243** [comment]
Sample screenshot if a suggestion is not compatible with language version of the project.
![Sharpen_Suggestion](https://user-images.githubusercontent.com/10205520/62361233-35633180-b538-11e9-923f-5dc70cba52a8.PNG)

**Comment by ironcev** [comment]
Thanks for digging deeper into the issue @shankyjain7243.

I'll answer the question from your PR #26 here so that we have all the things in one place and document at the same time some additional difficulties we have in this issue.

> We are using `CSharpLanguageVersions` class to identify language versions, which can also be done via `Microsoft.CodeAnalysis.CSharp.LanguageVersion` enum, do I replace the class with the enum around the source code ?

No, this replacement wouldn't work. Why? Because of these two major reasons:

- We need a richer abstraction of a C# version then just an enum. See #12. (I will try to add more info to the #12 so that is clear why we need a richer abstraction and what it should contain.)
- Even if a simple enum would be enough, we still cannot use the `Microsoft.CodeAnalysis.CSharp.LanguageVersion` enum. The reason lies in the problem described in the [Roslyn Dependencies at Build and Runtime](https://github.com/sharpenrocks/Sharpen/tree/master/lab/RoslynDependenciesAtBuildAndRuntime) lab experiment. Basically, we develop Sharpen with an older Roslyn version so that it can run in VS2017. In that version the `Microsoft.CodeAnalysis.CSharp.LanguageVersion` enum does not contain all the language version for which we have suggestions, e.g. the version 8.0.

In the meanwhile, two more difficulties popped up. The first one is explained in [this TODO](https://github.com/sharpenrocks/Sharpen/blob/90917b00866c1788bfb68839801e217a33b636ee/src/Sharpen.Engine/SharpenSuggestions/CSharp80/NullableReferenceTypes/Analyzers/EnableNullableContextAndDeclareIdentifierAsNullableAnalyzer.cs#L13). Basically, we can have analyzers that analyze files from a certain project and report findings in different projects. The solution provided in the PR does not support that, same as the existing implementation of the analyzers.

The second one is the larger-scale refactoring which we want to do, basically fixing all the points mentioned in the above TODO.

The PR #26 is a WIP as the title suggests but I wouldn't integrate it before we solve the other issues which are pretty heavy:

- [ ] Solving #12.
- [ ] Solving the [Roslyn Dependencies at Build and Runtime](https://github.com/sharpenrocks/Sharpen/tree/master/lab/RoslynDependenciesAtBuildAndRuntime) issue.
- [ ] Solving all the points in the [TODO](https://github.com/sharpenrocks/Sharpen/blob/90917b00866c1788bfb68839801e217a33b636ee/src/Sharpen.Engine/SharpenSuggestions/CSharp80/NullableReferenceTypes/Analyzers/EnableNullableContextAndDeclareIdentifierAsNullableAnalyzer.cs#L13).
- [ ] Solving the [connection between the suggestion and the framework and runtime version](https://github.com/sharpenrocks/Sharpen/issues/13#issuecomment-494343348).

So I will close the PR and I suggest that we postpone the work on this issue until all the points above are close. Thanks once again for providing more info!
